Firestore/core/src/firebase/firestore/local/local_serializer.cc

```diff
@@ -208,6 +211,123 @@ std::unique_ptr<model::NoDocument> LocalSerializer::DecodeNoDocument(
+  writer->WriteBytes(query_data.resume_token());
+
+  const Query& query = query_data.query();
+  if (query.IsDocumentQuery()) {
+    // TODO(rsgowman): Implement. Probably like this (once EncodeDocumentsTarget
+    // exists):
+    /*
+    writer->WriteTag({PB_WT_STRING, firestore_client_Target_documents_tag});
+    writer->WriteNestedMessage([&](Writer* writer) {
+      rpc_serializer_.EncodeDocumentsTarget(writer, query);
+    });
+    */
+    abort();
+  } else {
+    writer->WriteTag({PB_WT_STRING, firestore_client_Target_query_tag});
+    writer->WriteNestedMessage([&](Writer* writer) {
+      rpc_serializer_.EncodeQueryTarget(writer, query);
+    });
+  }
+}
+
+util::StatusOr<QueryData> LocalSerializer::DecodeQueryData(
+    const uint8_t* bytes, size_t length) const {
+  Reader reader = Reader::Wrap(bytes, length);
+  QueryData query_data = DecodeQueryData(&reader);
+  if (reader.status().ok()) {
+    return query_data;
+  } else {
+    return reader.status();
+  }
+}
+
+QueryData LocalSerializer::DecodeQueryData(Reader* reader) const {
+  if (!reader->status().ok()) return QueryData::Invalid();
+
+  int64_t target_id = 0;
+  SnapshotVersion version = SnapshotVersion::None();
+  std::vector<uint8_t> resume_token;
+  Query query = Query::Invalid();
+
+  while (reader->bytes_left()) {
+    Tag tag = reader->ReadTag();
+    if (!reader->status().ok()) return QueryData::Invalid();
+
+    switch (tag.field_number) {
+      case firestore_client_Target_target_id_tag:
+        if (!reader->RequireWireType(PB_WT_VARINT, tag))
+          return QueryData::Invalid();
+        target_id = reader->ReadInteger();
+        break;
+
+      case firestore_client_Target_snapshot_version_tag:
+        if (!reader->RequireWireType(PB_WT_STRING, tag))
+          return QueryData::Invalid();
+        version = SnapshotVersion{reader->ReadNestedMessage<Timestamp>(
+            rpc_serializer_.DecodeTimestamp)};
+        break;
+
+      case firestore_client_Target_resume_token_tag:
+        if (!reader->RequireWireType(PB_WT_STRING, tag))
+          return QueryData::Invalid();
+        resume_token = reader->ReadBytes();
+        break;
+
+      case firestore_client_Target_query_tag:
+        if (!reader->RequireWireType(PB_WT_STRING, tag))
+          return QueryData::Invalid();
+        // TODO(rsgowman): Clear 'documents' field (since query and documents
+        // are part of a 'oneof').
+        query =
+            reader->ReadNestedMessage<Query>(rpc_serializer_.DecodeQueryTarget);
+        break;
+
+      case firestore_client_Target_documents_tag:
+        if (!reader->RequireWireType(PB_WT_STRING, tag))
+          return QueryData::Invalid();
+        // Clear 'query' field (since query and documents are part of a 'oneof')
+        query = Query::Invalid();
+        // TODO(rsgowman): Implement.
+        abort();
+
+      default:
+        // Unknown tag. According to the proto spec, we need to ignore these.
+        reader->SkipField(tag);
```

Comment:
This bit of copypasta bugs me. Could we make `ReadTag` do this implicitly if called twice in succession?

Reply:
OK. I'm also in favor of dropping the comment. With the change to the current tag inside the reader we eliminate the `tag` parameter. Maybe this should be `reader->SkipUnknown()` for clarity?

Reply:
sgtm. I've added:

f) SkipField -> SkipUnknown() and eliminate repetitive comment

to the plan. (Though realistically, I'll lump this one together with the commit that eliminates `tag`.)